AsyncNats/NatsServerPool.cs

```diff
@@ -0,0 +1,176 @@
+﻿namespace EightyDecibel.AsyncNats
+{
+    using System;
+    using System.Buffers;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net;
+    using System.Threading.Tasks;
+    using Microsoft.Extensions.Logging;
+
+    internal class NatsServerPool
+    {
+
+        public List<NatsHost> Servers => new List<NatsHost>(_servers.Concat(_discoveredServers));
+
+        List<NatsHost> _servers = new List<NatsHost>();
+        List<NatsHost> _discoveredServers = new List<NatsHost>();
+        Random _random = new Random();
+        NatsHost? _lastSelectedServer = null;
+        INatsOptions _options;
+
+        object _sync = new object();
+
+        public NatsServerPool(INatsOptions options)
+        {
+
+#pragma warning disable CS0618 // Type or member is obsolete
+            //allow legacy server indication
+            if (options.Server != null && options.Servers==null)
+            {
+                options.Servers = new string[] { $"{options.Server.Address}:{options.Server.Port}" };
+            }
+#pragma warning restore CS0618 // Type or member is obsolete
+
+            var servers = options.Servers;
+
+            if(servers == null) throw new ArgumentNullException(nameof(servers));
+            if(servers.Count()==0) throw new ArgumentException("server list cannot be empty",nameof(servers));
+
+            _options = options;
+
+            foreach (var server in servers)
+                AddServer(_servers,server);
+        }
+
+        public NatsHost SelectServer(bool isRetry=false)
+        {
+            lock(_sync)
+            {
+                var combinedServers = _servers;
+                if(_options.ServersOptions.HasFlag(NatsServerPoolFlags.AllowDiscovery))
+                    combinedServers = _servers.Concat(_discoveredServers).ToList();
+
+                var selectedServer = combinedServers[0];
+
+                if (combinedServers.Count > 1)

```

Comment:
Invert this if statement:
````C#
if (combinedServers.Count == 1)
{
  _lastSelectedServer = selectedServer;
  return selectedServer!;
}
````
There will be less indenting in the rest of the function making it easier to read.

Reply:
Done

Reply:
Done.